File: bindings/r/R/safe_phone.R
# SPDX-License-Identifier: AGPL-3.0-or-later
# SPDX-FileCopyrightText: 2025 Hyperpolymath

#' Safe phone number validation, parsing, and formatting.
#'
#' Provides phone number parsing, validation, and E.164 formatting
#' with country code support.

#' Country calling codes and phone number patterns.
#'
#' Named list with country code, country name, and phone pattern info.
#' @export
country_codes <- list(
  # North America (NANP)
  US = list(code = "+1", name = "United States", pattern = "^[2-9][0-9]{9}$", length = 10L),
  CA = list(code = "+1", name = "Canada", pattern = "^[2-9][0-9]{9}$", length = 10L),
  # Europe
  GB = list(code = "+44", name = "United Kingdom", pattern = "^[1-9][0-9]{9,10}$", length = c(10L, 11L)),
  DE = list(code = "+49", name = "Germany", pattern = "^[1-9][0-9]{5,14}$", length = c(6L, 15L)),
  FR = list(code = "+33", name = "France", pattern = "^[1-9][0-9]{8}$", length = 9L),
  IT = list(code = "+39", name = "Italy", pattern = "^[0-9]{6,12}$", length = c(6L, 12L)),
  ES = list(code = "+34", name = "Spain", pattern = "^[6-9][0-9]{8}$", length = 9L),
  NL = list(code = "+31", name = "Netherlands", pattern = "^[1-9][0-9]{8}$", length = 9L),
  BE = list(code = "+32", name = "Belgium", pattern = "^[1-9][0-9]{7,8}$", length = c(8L, 9L)),
  CH = list(code = "+41", name = "Switzerland", pattern = "^[1-9][0-9]{8}$", length = 9L),
  AT = list(code = "+43", name = "Austria", pattern = "^[1-9][0-9]{3,12}$", length = c(4L, 13L)),
  SE = list(code = "+46", name = "Sweden", pattern = "^[1-9][0-9]{6,12}$", length = c(7L, 13L)),
  NO = list(code = "+47", name = "Norway", pattern = "^[2-9][0-9]{7}$", length = 8L),
  DK = list(code = "+45", name = "Denmark", pattern = "^[2-9][0-9]{7}$", length = 8L),
  FI = list(code = "+358", name = "Finland", pattern = "^[1-9][0-9]{4,11}$", length = c(5L, 12L)),
  PL = list(code = "+48", name = "Poland", pattern = "^[1-9][0-9]{8}$", length = 9L),
  PT = list(code = "+351", name = "Portugal", pattern = "^[2-9][0-9]{8}$", length = 9L),
  IE = list(code = "+353", name = "Ireland", pattern = "^[1-9][0-9]{6,9}$", length = c(7L, 10L)),
  # Asia Pacific
  CN = list(code = "+86", name = "China", pattern = "^1[0-9]{10}$", length = 11L),
  JP = list(code = "+81", name = "Japan", pattern = "^[1-9][0-9]{8,9}$", length = c(9L, 10L)),
  KR = list(code = "+82", name = "South Korea", pattern = "^[1-9][0-9]{7,10}$", length = c(8L, 11L)),
  IN = list(code = "+91", name = "India", pattern = "^[6-9][0-9]{9}$", length = 10L),
  AU = list(code = "+61", name = "Australia", pattern = "^[2-9][0-9]{8}$", length = 9L),
  NZ = list(code = "+64", name = "New Zealand", pattern = "^[2-9][0-9]{7,9}$", length = c(8L, 10L)),
  SG = list(code = "+65", name = "Singapore", pattern = "^[3689][0-9]{7}$", length = 8L),
  HK = list(code = "+852", name = "Hong Kong", pattern = "^[2-9][0-9]{7}$", length = 8L),
  TW = list(code = "+886", name = "Taiwan", pattern = "^[2-9][0-9]{7,8}$", length = c(8L, 9L)),
  PH = list(code = "+63", name = "Philippines", pattern = "^[2-9][0-9]{7,10}$", length = c(8L, 11L)),
  MY = list(code = "+60", name = "Malaysia", pattern = "^[1-9][0-9]{7,9}$", length = c(8L, 10L)),
  TH = list(code = "+66", name = "Thailand", pattern = "^[2-9][0-9]{7,8}$", length = c(8L, 9L)),
  ID = list(code = "+62", name = "Indonesia", pattern = "^[1-9][0-9]{7,12}$", length = c(8L, 13L)),
  VN = list(code = "+84", name = "Vietnam", pattern = "^[1-9][0-9]{8,9}$", length = c(9L, 10L)),
  # Middle East
  AE = list(code = "+971", name = "United Arab Emirates", pattern = "^[1-9][0-9]{8}$", length = 9L),
  SA = list(code = "+966", name = "Saudi Arabia", pattern = "^[1-9][0-9]{8}$", length = 9L),
  IL = list(code = "+972", name = "Israel", pattern = "^[2-9][0-9]{7,8}$", length = c(8L, 9L)),
  # Americas
  MX = list(code = "+52", name = "Mexico", pattern = "^[1-9][0-9]{9}$", length = 10L),
  BR = list(code = "+55", name = "Brazil", pattern = "^[1-9][0-9]{9,10}$", length = c(10L, 11L)),
  AR = list(code = "+54", name = "Argentina", pattern = "^[1-9][0-9]{9,10}$", length = c(10L, 11L)),
  CO = list(code = "+57", name = "Colombia", pattern = "^[1-9][0-9]{9}$", length = 10L),
  CL = list(code = "+56", name = "Chile", pattern = "^[2-9][0-9]{8}$", length = 9L),
  # Africa
  ZA = list(code = "+27", name = "South Africa", pattern = "^[1-9][0-9]{8}$", length = 9L),
  NG = list(code = "+234", name = "Nigeria", pattern = "^[7-9][0-9]{9}$", length = 10L),
  EG = list(code = "+20", name = "Egypt", pattern = "^[1-2][0-9]{8,9}$", length = c(9L, 10L)),
  KE = list(code = "+254", name = "Kenya", pattern = "^[1-9][0-9]{8}$", length = 9L)
)

#' Get country info by ISO country code.
#'
#' @param iso_code Two-letter ISO country code
#' @return Country info list or NULL
#' @export
get_country_info <- function(iso_code) {
  if (is.na(iso_code) || is.null(iso_code)) {
    return(NULL)
  }

  code <- toupper(iso_code)
  if (code %in% names(country_codes)) {
    country_codes[[code]]
  } else {
    NULL
  }
}

#' Get calling code for a country.
#'
#' @param iso_code Two-letter ISO country code
#' @return Calling code string or NA
#' @export
get_calling_code <- function(iso_code) {
  info <- get_country_info(iso_code)
  if (is.null(info)) {
    NA_character_
  } else {
    info$code
  }
}

#' Create a PhoneNumber object.
#'
#' @param national_number The national phone number (digits only)
#' @param country_code Two-letter ISO country code
#' @return PhoneNumber S3 object or NULL on invalid input
#' @keywords internal
new_phone_number <- function(national_number, country_code) {
  country_info <- get_country_info(country_code)
  if (is.null(country_info)) {
    return(NULL)
  }

  structure(
    list(
      national_number = national_number,
      country_code = toupper(country_code),
      calling_code = country_info$code
    ),
    class = "PhoneNumber"
  )
}

#' Check if object is a PhoneNumber.
#'
#' @param x Object to check
#' @return TRUE if x is a PhoneNumber object
#' @export
is_phone_number <- function(x) {
  inherits(x, "PhoneNumber")
}

#' Format PhoneNumber object.
#'
#' @param x PhoneNumber object
#' @param ... Additional arguments (ignored)
#' @return Formatted string
#' @export
format.PhoneNumber <- function(x, ...) {
  format_phone_e164(x)
}

#' Print PhoneNumber object.
#'
#' @param x PhoneNumber object
#' @param ... Additional arguments (ignored)
#' @export
print.PhoneNumber <- function(x, ...) {
  cat("PhoneNumber:", format_phone_e164(x), "\n")
  cat("  Country:", x$country_code, "\n")
  invisible(x)
}

#' Extract digits from a phone number string.
#'
#' @param phone_string Phone number string
#' @return String of digits only
#' @keywords internal
extract_digits <- function(phone_string) {
  gsub("[^0-9]", "", phone_string)
}

#' Parse a phone number string.
#'
#' @param phone_string Phone number string
#' @param default_country Default country code if not specified
#' @return PhoneNumber object or NULL on invalid input
#' @export
parse_phone <- function(phone_string, default_country = "US") {
  if (is.na(phone_string) || is.null(phone_string) || nchar(phone_string) == 0) {
    return(NULL)
  }

  # Check if it starts with + (international format)
  is_international <- startsWith(trimws(phone_string), "+")

  # Extract just the digits
  digits <- extract_digits(phone_string)

  if (nchar(digits) == 0) {
    return(NULL)
  }

  if (is_international) {
    # Try to match against known country codes
    for (country_name in names(country_codes)) {
      info <- country_codes[[country_name]]
      code_digits <- extract_digits(info$code)

      if (startsWith(digits, code_digits)) {
        national <- substr(digits, nchar(code_digits) + 1, nchar(digits))

        # Validate length
        expected_length <- info$length
        if (is.numeric(expected_length) && length(expected_length) == 1) {
          if (nchar(national) == expected_length) {
            return(new_phone_number(national, country_name))
          }
        } else if (length(expected_length) == 2) {
          if (nchar(national) >= expected_length[1] && nchar(national) <= expected_length[2]) {
            return(new_phone_number(national, country_name))
          }
        }
      }
    }

    return(NULL)
  } else {
    # Use default country
    country_info <- get_country_info(default_country)
    if (is.null(country_info)) {
      return(NULL)
    }

    # Remove leading 0 if present (common in many countries)
    if (startsWith(digits, "0")) {
      digits <- substr(digits, 2, nchar(digits))
    }

    # Validate length
    expected_length <- country_info$length
    if (is.numeric(expected_length) && length(expected_length) == 1) {
      if (nchar(digits) != expected_length) {
        return(NULL)
      }
    } else if (length(expected_length) == 2) {
      if (nchar(digits) < expected_length[1] || nchar(digits) > expected_length[2]) {
        return(NULL)
      }
    }

    new_phone_number(digits, default_country)
  }
}

#' Format phone number in E.164 format.
#'
#' @param phone PhoneNumber object or string
#' @param default_country Default country for string input
#' @return E.164 formatted string or NA
#' @export
format_phone_e164 <- function(phone, default_country = "US") {
  if (is.character(phone)) {
    phone <- parse_phone(phone, default_country)
  }

  if (is.null(phone) || !is_phone_number(phone)) {
    return(NA_character_)
  }

  paste0(phone$calling_code, phone$national_number)
}

#' Format phone number for display (national format).
#'
#' @param phone PhoneNumber object or string
#' @param default_country Default country for string input
#' @return Formatted string or NA
#' @export
format_phone_national <- function(phone, default_country = "US") {
  if (is.character(phone)) {
    phone <- parse_phone(phone, default_country)
  }

  if (is.null(phone) || !is_phone_number(phone)) {
    return(NA_character_)
  }

  digits <- phone$national_number

  # US/CA formatting
  if (phone$country_code %in% c("US", "CA") && nchar(digits) == 10) {
    return(paste0(
      "(", substr(digits, 1, 3), ") ",
      substr(digits, 4, 6), "-",
      substr(digits, 7, 10)
    ))
  }

  # UK formatting
  if (phone$country_code == "GB") {
    if (nchar(digits) == 10) {
      return(paste0(
        substr(digits, 1, 4), " ",
        substr(digits, 5, 7), " ",
        substr(digits, 8, 10)
      ))
    } else if (nchar(digits) == 11) {
      return(paste0(
        substr(digits, 1, 5), " ",
        substr(digits, 6, 8), " ",
        substr(digits, 9, 11)
      ))
    }
  }

  # Default: just return the national number
  digits
}

#' Format phone number for display (international format).
#'
#' @param phone PhoneNumber object or string
#' @param default_country Default country for string input
#' @return Formatted string or NA
#' @export
format_phone_international <- function(phone, default_country = "US") {
  if (is.character(phone)) {
    phone <- parse_phone(phone, default_country)
  }

  if (is.null(phone) || !is_phone_number(phone)) {
    return(NA_character_)
  }

  paste0(phone$calling_code, " ", format_phone_national(phone))
}

#' Validate a phone number string.
#'
#' @param phone_string Phone number string
#' @param country Expected country code (optional)
#' @return TRUE if valid
#' @export
is_valid_phone <- function(phone_string, country = NULL) {
  if (is.null(country)) {
    # Try to parse as international
    phone <- parse_phone(phone_string, "US")
    return(!is.null(phone))
  }

  phone <- parse_phone(phone_string, country)

  if (is.null(phone)) {
    return(FALSE)
  }

  # Verify country matches
  toupper(country) == phone$country_code
}

#' Get the country of a phone number.
#'
#' @param phone PhoneNumber object or string
#' @return Country code or NA
#' @export
phone_country <- function(phone) {
  if (is.character(phone)) {
    phone <- parse_phone(phone, "US")
  }

  if (is.null(phone) || !is_phone_number(phone)) {
    return(NA_character_)
  }

  phone$country_code
}

#' Get the national number of a phone.
#'
#' @param phone PhoneNumber object or string
#' @return National number string or NA
#' @export
phone_national_number <- function(phone) {
  if (is.character(phone)) {
    phone <- parse_phone(phone, "US")
  }

  if (is.null(phone) || !is_phone_number(phone)) {
    return(NA_character_)
  }

  phone$national_number
}

#' Compare two phone numbers for equality.
#'
#' @param phone1 First phone (object or string)
#' @param phone2 Second phone (object or string)
#' @return TRUE if equal
#' @export
phones_equal <- function(phone1, phone2) {
  if (is.character(phone1)) {
    phone1 <- parse_phone(phone1, "US")
  }
  if (is.character(phone2)) {
    phone2 <- parse_phone(phone2, "US")
  }

  if (is.null(phone1) || is.null(phone2)) {
    return(FALSE)
  }

  format_phone_e164(phone1) == format_phone_e164(phone2)
}

#' Equality operator for PhoneNumber objects.
#'
#' @param e1 First PhoneNumber
#' @param e2 Second PhoneNumber
#' @return TRUE if equal
#' @export
`==.PhoneNumber` <- function(e1, e2) {
  phones_equal(e1, e2)
}

#' Mask a phone number for display.
#'
#' @param phone PhoneNumber object or string
#' @param visible_digits Number of digits to show at end (default 4)
#' @return Masked phone string or NA
#' @export
mask_phone <- function(phone, visible_digits = 4L) {
  if (is.character(phone)) {
    phone <- parse_phone(phone, "US")
  }

  if (is.null(phone) || !is_phone_number(phone)) {
    return(NA_character_)
  }

  national <- phone$national_number
  len <- nchar(national)

  if (len <= visible_digits) {
    return(paste0(phone$calling_code, " ", national))
  }

  masked_part <- paste0(rep("*", len - visible_digits), collapse = "")
  visible_part <- substr(national, len - visible_digits + 1, len)

  paste0(phone$calling_code, " ", masked_part, visible_part)
}

#' Check if phone number is a mobile number.
#'
#' This is a heuristic based on common mobile prefixes.
#'
#' @param phone PhoneNumber object or string
#' @return TRUE if likely mobile, NA if unknown
#' @export
is_mobile_phone <- function(phone) {
  if (is.character(phone)) {
    phone <- parse_phone(phone, "US")
  }

  if (is.null(phone) || !is_phone_number(phone)) {
    return(NA)
  }

  national <- phone$national_number

  # Country-specific mobile detection
  switch(phone$country_code,
    # US/CA: All area codes can be mobile
    "US" = TRUE,
    "CA" = TRUE,
    # UK: Mobile starts with 7
    "GB" = startsWith(national, "7"),
    # Germany: Mobile starts with 15, 16, 17
    "DE" = grepl("^1[567]", national),
    # France: Mobile starts with 6 or 7
    "FR" = grepl("^[67]", national),
    # China: Mobile starts with 1
    "CN" = startsWith(national, "1"),
    # India: Mobile starts with 6-9
    "IN" = grepl("^[6-9]", national),
    # Australia: Mobile starts with 4
    "AU" = startsWith(national, "4"),
    # Unknown
    NA
  )
}

File: bindings/ocaml/lib/safe_phone.ml
(* SPDX-License-Identifier: Apache-2.0 *)
(* SPDX-FileCopyrightText: 2025 Hyperpolymath *)

(** Safe phone number validation and formatting. *)

(** Country calling codes per ITU-T E.164. *)
type country_code =
  | US  (** +1 United States *)
  | CA  (** +1 Canada *)
  | GB  (** +44 United Kingdom *)
  | DE  (** +49 Germany *)
  | FR  (** +33 France *)
  | IT  (** +39 Italy *)
  | ES  (** +34 Spain *)
  | JP  (** +81 Japan *)
  | CN  (** +86 China *)
  | IN  (** +91 India *)
  | AU  (** +61 Australia *)
  | BR  (** +55 Brazil *)
  | MX  (** +52 Mexico *)
  | KR  (** +82 South Korea *)
  | RU  (** +7 Russia *)
  | NL  (** +31 Netherlands *)
  | SE  (** +46 Sweden *)
  | CH  (** +41 Switzerland *)
  | BE  (** +32 Belgium *)
  | AT  (** +43 Austria *)
  | PL  (** +48 Poland *)
  | Other of int  (** Other country code *)

(** Phone number representation. *)
type t = {
  country: country_code;
  national_number: string;  (** National significant number (digits only) *)
}

type error =
  | Invalid_format
  | Invalid_country_code
  | Invalid_length
  | Invalid_characters

(** Get numeric calling code for country. *)
let calling_code_of_country = function
  | US | CA -> 1
  | GB -> 44
  | DE -> 49
  | FR -> 33
  | IT -> 39
  | ES -> 34
  | JP -> 81
  | CN -> 86
  | IN -> 91
  | AU -> 61
  | BR -> 55
  | MX -> 52
  | KR -> 82
  | RU -> 7
  | NL -> 31
  | SE -> 46
  | CH -> 41
  | BE -> 32
  | AT -> 43
  | PL -> 48
  | Other n -> n

(** Get country from numeric calling code. *)
let country_of_calling_code = function
  | 1 -> US  (* Note: Both US and CA use +1, defaults to US *)
  | 44 -> GB
  | 49 -> DE
  | 33 -> FR
  | 39 -> IT
  | 34 -> ES
  | 81 -> JP
  | 86 -> CN
  | 91 -> IN
  | 61 -> AU
  | 55 -> BR
  | 52 -> MX
  | 82 -> KR
  | 7 -> RU
  | 31 -> NL
  | 46 -> SE
  | 41 -> CH
  | 32 -> BE
  | 43 -> AT
  | 48 -> PL
  | n -> Other n

(** Country code to ISO 3166-1 alpha-2 string. *)
let country_to_string = function
  | US -> "US" | CA -> "CA" | GB -> "GB" | DE -> "DE" | FR -> "FR"
  | IT -> "IT" | ES -> "ES" | JP -> "JP" | CN -> "CN" | IN -> "IN"
  | AU -> "AU" | BR -> "BR" | MX -> "MX" | KR -> "KR" | RU -> "RU"
  | NL -> "NL" | SE -> "SE" | CH -> "CH" | BE -> "BE" | AT -> "AT"
  | PL -> "PL" | Other n -> Printf.sprintf "+%d" n

(** Extract only digit characters from a string. *)
let extract_digits str =
  let buf = Buffer.create (String.length str) in
  String.iter (fun c ->
    if c >= '0' && c <= '9' then Buffer.add_char buf c
  ) str;
  Buffer.contents buf

(** Validate that string contains only valid phone characters. *)
let has_valid_chars str =
  String.for_all (fun c ->
    (c >= '0' && c <= '9') || c = '+' || c = '-' || c = ' ' ||
    c = '(' || c = ')' || c = '.'
  ) str

(** Get expected national number length range for country. *)
let national_number_length_range = function
  | US | CA -> (10, 10)
  | GB -> (10, 11)
  | DE -> (6, 13)
  | FR -> (9, 9)
  | IT -> (6, 11)
  | ES -> (9, 9)
  | JP -> (9, 10)
  | CN -> (11, 11)
  | IN -> (10, 10)
  | AU -> (9, 9)
  | BR -> (10, 11)
  | MX -> (10, 10)
  | KR -> (8, 11)
  | RU -> (10, 10)
  | NL -> (9, 9)
  | SE -> (7, 13)
  | CH -> (9, 9)
  | BE -> (8, 9)
  | AT -> (4, 13)
  | PL -> (9, 9)
  | Other _ -> (4, 15)  (* E.164 allows 4-15 digits for national number *)

(** Parse phone number from E.164 format (+1234567890). *)
let parse_e164 str =
  if String.length str < 2 then Error Invalid_format
  else if str.[0] <> '+' then Error Invalid_format
  else
    let digits = extract_digits str in
    if String.length digits < 7 || String.length digits > 15 then
      Error Invalid_length
    else
      (* Try to match country codes (1-3 digits) *)
      let try_code len =
        if len > String.length digits then None
        else
          let code_str = String.sub digits 0 len in
          try
            let code = int_of_string code_str in
            let country = country_of_calling_code code in
            let national = String.sub digits len (String.length digits - len) in
            let min_len, max_len = national_number_length_range country in
            if String.length national >= min_len && String.length national <= max_len then
              Some { country; national_number = national }
            else
              None
          with _ -> None
      in
      (* Try 1, 2, then 3 digit codes *)
      match try_code 1 with
      | Some phone -> Ok phone
      | None ->
          match try_code 2 with
          | Some phone -> Ok phone
          | None ->
              match try_code 3 with
              | Some phone -> Ok phone
              | None -> Error Invalid_country_code

(** Parse phone number with explicit country. *)
let parse ?(country=US) str =
  if not (has_valid_chars str) then Error Invalid_characters
  else
    let digits = extract_digits str in
    let min_len, max_len = national_number_length_range country in
    if String.length digits < min_len || String.length digits > max_len then
      Error Invalid_length
    else
      Ok { country; national_number = digits }

(** Format to E.164 international format. *)
let format_e164 phone =
  Printf.sprintf "+%d%s"
    (calling_code_of_country phone.country)
    phone.national_number

(** Format for display (country-specific). *)
let format phone =
  let n = phone.national_number in
  match phone.country with
  | US | CA when String.length n = 10 ->
      Printf.sprintf "(%s) %s-%s"
        (String.sub n 0 3) (String.sub n 3 3) (String.sub n 6 4)
  | GB when String.length n >= 10 ->
      Printf.sprintf "+44 %s %s"
        (String.sub n 0 4) (String.sub n 4 (String.length n - 4))
  | DE ->
      Printf.sprintf "+49 %s" n
  | FR when String.length n = 9 ->
      Printf.sprintf "+33 %s %s %s %s %s"
        (String.sub n 0 1) (String.sub n 1 2) (String.sub n 3 2)
        (String.sub n 5 2) (String.sub n 7 2)
  | _ ->
      format_e164 phone

(** Format as national number only. *)
let format_national phone =
  phone.national_number

(** Check if phone number is valid. *)
let is_valid phone =
  let min_len, max_len = national_number_length_range phone.country in
  let len = String.length phone.national_number in
  len >= min_len && len <= max_len &&
  String.for_all (fun c -> c >= '0' && c <= '9') phone.national_number

(** Get the country code. *)
let get_country phone = phone.country

(** Get the calling code as integer. *)
let get_calling_code phone = calling_code_of_country phone.country

(** Get the national number. *)
let get_national_number phone = phone.national_number

(** Compare two phone numbers. *)
let compare a b =
  let cmp = compare (calling_code_of_country a.country) (calling_code_of_country b.country) in
  if cmp <> 0 then cmp
  else String.compare a.national_number b.national_number

(** Check equality of two phone numbers. *)
let equal a b =
  compare a b = 0

(** Normalize phone number (strip formatting, standardize). *)
let normalize phone =
  { phone with national_number = extract_digits phone.national_number }

(** Check if number is a mobile number (heuristic, country-specific). *)
let is_mobile phone =
  let n = phone.national_number in
  if String.length n = 0 then false
  else
    match phone.country with
    | US | CA ->
        (* US/CA: All 10-digit numbers can be mobile *)
        String.length n = 10
    | GB ->
        (* UK mobile: starts with 7 *)
        n.[0] = '7'
    | DE ->
        (* German mobile: starts with 15, 16, 17 *)
        String.length n >= 2 && n.[0] = '1' &&
        (n.[1] = '5' || n.[1] = '6' || n.[1] = '7')
    | FR ->
        (* French mobile: starts with 6 or 7 *)
        n.[0] = '6' || n.[0] = '7'
    | _ ->
        (* Default: cannot determine *)
        false
